File: 0103.BinaryTreeZigzagLevelOrderTraversal.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

 var ans [][]int

 func zigzagLevelOrder(root *TreeNode) [][]int {
     ans = [][]int{}
     dfs(root, 0)
     for i := 1; i < len(ans); i += 2 {
         reverse(ans[i])
     }
     return ans
 }
 
 func dfs(node *TreeNode, i int) {
     if node == nil {
         return
     }
     if len(ans) == i {
         ans = append(ans, []int{node.Val})
     } else {
         ans[i] = append(ans[i], node.Val)
     }
     dfs(node.Left, i+1)
     dfs(node.Right, i+1)
 }
 
 func reverse(nums []int) {
     for l, r := 0, len(nums)-1; l < r; l, r = l+1, r-1 {
         nums[l], nums[r] = nums[r], nums[l]
     }
 }

File: 0000-0999/0103.BinaryTreeZigzagLevelOrderTraversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def zigzagLevelOrder(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        if not root:
            return []
        level = [root]
        ans = []
        flag = True
        while level:
            temp = [node.val for node in level]
            level = [child for f in level for child in (f.left, f.right) if child]
            if not flag:
                temp.reverse()
            ans.append(temp)
            flag = not flag
        return ans
